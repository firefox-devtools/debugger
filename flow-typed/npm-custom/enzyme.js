/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// Using this custom definition of the enzyme module avoids generating some
// flow errors that are unrelated to uses of enzyme itself...
// It would be nice to clean this up.
declare module "enzyme" {
  declare module.exports: any;
}
