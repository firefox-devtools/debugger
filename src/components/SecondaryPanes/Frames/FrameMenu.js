// @flow
import { showMenu } from "devtools-launchpad";
import { copyToTheClipboard } from "../../../utils/clipboard";
import type { LocalFrame } from "./types";
import type { ContextMenuItem } from "debugger-html";
import { kebabCase } from "lodash";

const blackboxString = "sourceFooter.blackbox";
const unblackboxString = "sourceFooter.unblackbox";

function formatMenuElement(
  labelString: string,
  click: Function,
  disabled: boolean = false
): ContextMenuItem {
  const label = L10N.getStr(labelString);
  const accesskey = L10N.getStr(`${labelString}.accesskey`);
  const id = `node-menu-${kebabCase(label)}`;
  return {
    id,
    label,
    accesskey,
    disabled,
    click
  };
}

function copySourceElement(url) {
  return formatMenuElement("copySourceUri2", () => copyToTheClipboard(url));
}

function copyStackTraceElement(copyStackTrace) {
  return formatMenuElement("copyStackTrace", () => copyStackTrace());
}

function toggleFrameworkGroupingElement(
  toggleFrameworkGrouping,
  frameworkGroupingOn
) {
  const actionType = frameworkGroupingOn
    ? "framework.disableGrouping"
    : "framework.enableGrouping";

  return formatMenuElement(actionType, () => toggleFrameworkGrouping());
}

function blackBoxSource(source, toggleBlackBox) {
  const toggleBlackBoxString = source.isBlackBoxed
    ? unblackboxString
    : blackboxString;

  return formatMenuElement(toggleBlackBoxString, () => toggleBlackBox(source));
}

export default function FrameMenu(
  frame: LocalFrame,
  frameworkGroupingOn: boolean,
  callbacks: Object,
  event: SyntheticKeyboardEvent
) {
  event.stopPropagation();
  event.preventDefault();

  const menuOptions = [];

  const source = frame.source;

  const toggleFrameworkElement = toggleFrameworkGroupingElement(
    callbacks.toggleFrameworkGrouping,
    frameworkGroupingOn
  );
  menuOptions.push(toggleFrameworkElement);

  if (source) {
    const copySourceUri2 = copySourceElement(source.url);
    menuOptions.push(copySourceUri2);
    menuOptions.push(blackBoxSource(source, callbacks.toggleBlackBox));
  }

  const copyStackTraceItem = copyStackTraceElement(callbacks.copyStackTrace);

  menuOptions.push(copyStackTraceItem);

  showMenu(event, menuOptions);
}
