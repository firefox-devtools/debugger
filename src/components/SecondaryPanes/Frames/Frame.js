/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow
import React, { Component } from "react";
import PropTypes from "prop-types";

import classNames from "classnames";
import Svg from "../../shared/Svg";

import { formatDisplayName } from "../../../utils/pause/frames";
import { getFilename, getFileURL } from "../../../utils/source";
import FrameMenu from "./FrameMenu";

import type { Frame } from "../../../types";
import type { LocalFrame } from "./types";

type FrameTitleProps = {
  frame: Frame,
  options: Object,
  l10n: Object
};

function FrameTitle({ frame, options = {}, l10n }: FrameTitleProps) {
  const displayName = formatDisplayName(frame, options, l10n);
  return <span className="title">{displayName}</span>;
}

type FrameLocationProps = { frame: LocalFrame, displayFullUrl: boolean };

function FrameLocation({ frame, displayFullUrl = false }: FrameLocationProps) {
  if (!frame.source) {
    return null;
  }

  if (frame.library) {
    return (
      <span className="location">
        {frame.library}
        <Svg name={frame.library.toLowerCase()} className="annotation-logo" />
      </span>
    );
  }

  const { location, source } = frame;
  const filename = displayFullUrl
    ? getFileURL(source, false)
    : getFilename(source);

  return (
    <span className="location">
      <span className="filename">{filename}</span>:
      <span className="line">{location.line}</span>
    </span>
  );
}

FrameLocation.displayName = "FrameLocation";

type FrameComponentProps = {
  frame: LocalFrame,
  selectedFrame: LocalFrame,
  copyStackTrace: Function,
  toggleFrameworkGrouping: Function,
  selectFrame: Function,
  frameworkGroupingOn: boolean,
  hideLocation: boolean,
  shouldMapDisplayName: boolean,
  toggleBlackBox: Function,
  displayFullUrl: boolean,
  getFrameTitle?: string => string,
  disableContextMenu: boolean
};

export default class FrameComponent extends Component<FrameComponentProps> {
  static defaultProps = {
    hideLocation: false,
    shouldMapDisplayName: true,
    disableContextMenu: false
  };

  onContextMenu(event: SyntheticMouseEvent<HTMLElement>) {
    const {
      frame,
      copyStackTrace,
      toggleFrameworkGrouping,
      toggleBlackBox,
      frameworkGroupingOn
    } = this.props;
    FrameMenu(
      frame,
      frameworkGroupingOn,
      { copyStackTrace, toggleFrameworkGrouping, toggleBlackBox },
      event
    );
  }

  onMouseDown(
    e: SyntheticMouseEvent<HTMLElement>,
    frame: Frame,
    selectedFrame: Frame
  ) {
    if (e.button !== 0) {
      return;
    }
    this.props.selectFrame(frame);
  }

  onKeyUp(
    event: SyntheticKeyboardEvent<HTMLElement>,
    frame: Frame,
    selectedFrame: Frame
  ) {
    if (event.key != "Enter") {
      return;
    }
    this.props.selectFrame(frame);
  }

  render() {
    const {
      frame,
      selectedFrame,
      hideLocation,
      shouldMapDisplayName,
      displayFullUrl,
      getFrameTitle,
      disableContextMenu
    } = this.props;
    const { l10n } = this.context;

    const className = classNames("frame", {
      selected: selectedFrame && selectedFrame.id === frame.id
    });

    const title = getFrameTitle
      ? getFrameTitle(
          `${getFileURL(frame.source, false)}:${frame.location.line}`
        )
      : undefined;

    const tabChar = "\t";
    const newLineChar = "\n";

    return (
      <li
        key={frame.id}
        className={className}
        onMouseDown={e => this.onMouseDown(e, frame, selectedFrame)}
        onKeyUp={e => this.onKeyUp(e, frame, selectedFrame)}
        onContextMenu={disableContextMenu ? null : e => this.onContextMenu(e)}
        tabIndex={0}
        title={title}
      >
        {tabChar}
        <FrameTitle
          frame={frame}
          options={{ shouldMapDisplayName }}
          l10n={l10n}
        />
        {!hideLocation && " "}
        {!hideLocation && (
          <FrameLocation frame={frame} displayFullUrl={displayFullUrl} />
        )}
        {newLineChar}
      </li>
    );
  }
}

FrameComponent.displayName = "Frame";
FrameComponent.contextTypes = { l10n: PropTypes.object };
