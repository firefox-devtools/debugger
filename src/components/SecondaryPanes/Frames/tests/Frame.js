import React from "react";
import { shallow } from "enzyme";
import Frame from "../Frame.js";

import FrameMenu from "../FrameMenu";
jest.mock("../FrameMenu", () => jest.fn());

function render(frameToSelect = {}, overrides = {}) {
  const defaultFrame = {
    id: 1,
    source: {
      url: "foo-view.js",
      isBlackBoxed: false
    },
    displayName: "renderFoo",
    frameworkGroupingOn: false,
    toggleFrameworkGrouping: jest.fn(),
    library: false,
    location: {
      line: 10
    }
  };
  const frame = Object.assign({}, defaultFrame, overrides);
  const selectedFrame = Object.assign({}, frame, frameToSelect);
  const selectFrame = jest.fn();
  const toggleBlackBox = jest.fn();

  const props = {
    frame,
    selectedFrame,
    copyStackTrace: jest.fn(),
    contextTypes: {},
    selectFrame,
    toggleBlackBox
  };
  const component = shallow(<Frame {...props} />);
  return { component, props };
}

describe("Frame", () => {
  it("user frame", () => {
    const { component } = render();
    expect(component).toMatchSnapshot();
  });

  it("user frame (not selected)", () => {
    const { component } = render({ id: 2 });
    expect(component).toMatchSnapshot();
  });

  it("library frame", () => {
    const backboneFrame = {
      id: 3,
      source: { url: "backbone.js" },
      displayName: "updateEvents",
      library: "backbone",
      location: {
        line: 12
      }
    };

    const { component } = render({ id: 3 }, backboneFrame);
    expect(component).toMatchSnapshot();
  });

  describe("mouse events", () => {
    it("calls FrameMenu on right click", () => {
      const { component, props } = render();
      const { copyStackTrace, toggleFrameworkGrouping, toggleBlackBox } = props;
      const mockEvent = "mockEvent";
      component.simulate("contextmenu", mockEvent);

      expect(FrameMenu).toHaveBeenCalledWith(
        props.frame,
        props.frameworkGroupingOn,
        {
          copyStackTrace,
          toggleFrameworkGrouping,
          toggleBlackBox
        },
        mockEvent
      );
    });
  });
});
