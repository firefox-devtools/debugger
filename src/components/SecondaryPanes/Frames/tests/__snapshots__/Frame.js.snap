// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frame library frame 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li
    className="frame selected"
    onContextMenu={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
>
    <div
        className="title"
    >
        updateEvents
    </div>
    <div
        className="location"
    >
        backbone
        <InlineSVG
            className="backbone annotation-logo"
            element="i"
            raw={false}
            src="<svg></svg>"
        />
    </div>
</li>,
  "nodes": Array [
    <li
      className="frame selected"
      onContextMenu={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      tabIndex={0}
>
      <div
            className="title"
      >
            updateEvents
      </div>
      <div
            className="location"
      >
            backbone
            <InlineSVG
                  className="backbone annotation-logo"
                  element="i"
                  raw={false}
                  src="<svg></svg>"
            />
      </div>
</li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Frame
        contextTypes={Object {}}
        copyStackTrace={[Function]}
        frame={
                Object {
                        "displayName": "updateEvents",
                        "frameworkGroupingOn": false,
                        "id": 3,
                        "library": "backbone",
                        "location": Object {
                          "line": 12,
                        },
                        "source": Object {
                          "url": "backbone.js",
                        },
                        "toggleFrameworkGrouping": [Function],
                      }
        }
        hideLocation={false}
        selectFrame={[Function]}
        selectedFrame={
                Object {
                        "displayName": "updateEvents",
                        "frameworkGroupingOn": false,
                        "id": 3,
                        "library": "backbone",
                        "location": Object {
                          "line": 12,
                        },
                        "source": Object {
                          "url": "backbone.js",
                        },
                        "toggleFrameworkGrouping": [Function],
                      }
        }
        shouldMapDisplayName={true}
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": FrameComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "contextTypes": Object {},
          "copyStackTrace": [Function],
          "frame": Object {
            "displayName": "updateEvents",
            "frameworkGroupingOn": false,
            "id": 3,
            "library": "backbone",
            "location": Object {
              "line": 12,
            },
            "source": Object {
              "url": "backbone.js",
            },
            "toggleFrameworkGrouping": [Function],
          },
          "hideLocation": false,
          "selectFrame": [Function],
          "selectedFrame": Object {
            "displayName": "updateEvents",
            "frameworkGroupingOn": false,
            "id": 3,
            "library": "backbone",
            "location": Object {
              "line": 12,
            },
            "source": Object {
              "url": "backbone.js",
            },
            "toggleFrameworkGrouping": [Function],
          },
          "shouldMapDisplayName": true,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <li
          className="frame selected"
          onContextMenu={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
>
          <div
                    className="title"
          >
                    updateEvents
          </div>
          <div
                    className="location"
          >
                    backbone
                    <InlineSVG
                              className="backbone annotation-logo"
                              element="i"
                              raw={false}
                              src="<svg></svg>"
                    />
          </div>
</li>,
        "_debugID": 6,
        "_renderedOutput": <li
          className="frame selected"
          onContextMenu={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
>
          <div
                    className="title"
          >
                    updateEvents
          </div>
          <div
                    className="location"
          >
                    backbone
                    <InlineSVG
                              className="backbone annotation-logo"
                              element="i"
                              raw={false}
                              src="<svg></svg>"
                    />
          </div>
</li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Frame
    contextTypes={Object {}}
    copyStackTrace={[Function]}
    frame={
        Object {
            "displayName": "updateEvents",
            "frameworkGroupingOn": false,
            "id": 3,
            "library": "backbone",
            "location": Object {
              "line": 12,
            },
            "source": Object {
              "url": "backbone.js",
            },
            "toggleFrameworkGrouping": [Function],
          }
    }
    hideLocation={false}
    selectFrame={[Function]}
    selectedFrame={
        Object {
            "displayName": "updateEvents",
            "frameworkGroupingOn": false,
            "id": 3,
            "library": "backbone",
            "location": Object {
              "line": 12,
            },
            "source": Object {
              "url": "backbone.js",
            },
            "toggleFrameworkGrouping": [Function],
          }
    }
    shouldMapDisplayName={true}
/>,
}
`;

exports[`Frame user frame (not selected) 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li
    className="frame"
    onContextMenu={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
>
    <div
        className="title"
    >
        renderFoo
    </div>
    <div
        className="location"
    >
        foo-view.js: 10
    </div>
</li>,
  "nodes": Array [
    <li
      className="frame"
      onContextMenu={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      tabIndex={0}
>
      <div
            className="title"
      >
            renderFoo
      </div>
      <div
            className="location"
      >
            foo-view.js: 10
      </div>
</li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Frame
        contextTypes={Object {}}
        copyStackTrace={[Function]}
        frame={
                Object {
                        "displayName": "renderFoo",
                        "frameworkGroupingOn": false,
                        "id": 1,
                        "library": false,
                        "location": Object {
                          "line": 10,
                        },
                        "source": Object {
                          "url": "foo-view.js",
                        },
                        "toggleFrameworkGrouping": [Function],
                      }
        }
        hideLocation={false}
        selectFrame={[Function]}
        selectedFrame={
                Object {
                        "displayName": "renderFoo",
                        "frameworkGroupingOn": false,
                        "id": 2,
                        "library": false,
                        "location": Object {
                          "line": 10,
                        },
                        "source": Object {
                          "url": "foo-view.js",
                        },
                        "toggleFrameworkGrouping": [Function],
                      }
        }
        shouldMapDisplayName={true}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": FrameComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "contextTypes": Object {},
          "copyStackTrace": [Function],
          "frame": Object {
            "displayName": "renderFoo",
            "frameworkGroupingOn": false,
            "id": 1,
            "library": false,
            "location": Object {
              "line": 10,
            },
            "source": Object {
              "url": "foo-view.js",
            },
            "toggleFrameworkGrouping": [Function],
          },
          "hideLocation": false,
          "selectFrame": [Function],
          "selectedFrame": Object {
            "displayName": "renderFoo",
            "frameworkGroupingOn": false,
            "id": 2,
            "library": false,
            "location": Object {
              "line": 10,
            },
            "source": Object {
              "url": "foo-view.js",
            },
            "toggleFrameworkGrouping": [Function],
          },
          "shouldMapDisplayName": true,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <li
          className="frame"
          onContextMenu={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
>
          <div
                    className="title"
          >
                    renderFoo
          </div>
          <div
                    className="location"
          >
                    foo-view.js: 10
          </div>
</li>,
        "_debugID": 4,
        "_renderedOutput": <li
          className="frame"
          onContextMenu={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
>
          <div
                    className="title"
          >
                    renderFoo
          </div>
          <div
                    className="location"
          >
                    foo-view.js: 10
          </div>
</li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Frame
    contextTypes={Object {}}
    copyStackTrace={[Function]}
    frame={
        Object {
            "displayName": "renderFoo",
            "frameworkGroupingOn": false,
            "id": 1,
            "library": false,
            "location": Object {
              "line": 10,
            },
            "source": Object {
              "url": "foo-view.js",
            },
            "toggleFrameworkGrouping": [Function],
          }
    }
    hideLocation={false}
    selectFrame={[Function]}
    selectedFrame={
        Object {
            "displayName": "renderFoo",
            "frameworkGroupingOn": false,
            "id": 2,
            "library": false,
            "location": Object {
              "line": 10,
            },
            "source": Object {
              "url": "foo-view.js",
            },
            "toggleFrameworkGrouping": [Function],
          }
    }
    shouldMapDisplayName={true}
/>,
}
`;

exports[`Frame user frame 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li
    className="frame selected"
    onContextMenu={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
>
    <div
        className="title"
    >
        renderFoo
    </div>
    <div
        className="location"
    >
        foo-view.js: 10
    </div>
</li>,
  "nodes": Array [
    <li
      className="frame selected"
      onContextMenu={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      tabIndex={0}
>
      <div
            className="title"
      >
            renderFoo
      </div>
      <div
            className="location"
      >
            foo-view.js: 10
      </div>
</li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Frame
        contextTypes={Object {}}
        copyStackTrace={[Function]}
        frame={
                Object {
                        "displayName": "renderFoo",
                        "frameworkGroupingOn": false,
                        "id": 1,
                        "library": false,
                        "location": Object {
                          "line": 10,
                        },
                        "source": Object {
                          "url": "foo-view.js",
                        },
                        "toggleFrameworkGrouping": [Function],
                      }
        }
        hideLocation={false}
        selectFrame={[Function]}
        selectedFrame={
                Object {
                        "displayName": "renderFoo",
                        "frameworkGroupingOn": false,
                        "id": 1,
                        "library": false,
                        "location": Object {
                          "line": 10,
                        },
                        "source": Object {
                          "url": "foo-view.js",
                        },
                        "toggleFrameworkGrouping": [Function],
                      }
        }
        shouldMapDisplayName={true}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": FrameComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "contextTypes": Object {},
          "copyStackTrace": [Function],
          "frame": Object {
            "displayName": "renderFoo",
            "frameworkGroupingOn": false,
            "id": 1,
            "library": false,
            "location": Object {
              "line": 10,
            },
            "source": Object {
              "url": "foo-view.js",
            },
            "toggleFrameworkGrouping": [Function],
          },
          "hideLocation": false,
          "selectFrame": [Function],
          "selectedFrame": Object {
            "displayName": "renderFoo",
            "frameworkGroupingOn": false,
            "id": 1,
            "library": false,
            "location": Object {
              "line": 10,
            },
            "source": Object {
              "url": "foo-view.js",
            },
            "toggleFrameworkGrouping": [Function],
          },
          "shouldMapDisplayName": true,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <li
          className="frame selected"
          onContextMenu={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
>
          <div
                    className="title"
          >
                    renderFoo
          </div>
          <div
                    className="location"
          >
                    foo-view.js: 10
          </div>
</li>,
        "_debugID": 2,
        "_renderedOutput": <li
          className="frame selected"
          onContextMenu={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
>
          <div
                    className="title"
          >
                    renderFoo
          </div>
          <div
                    className="location"
          >
                    foo-view.js: 10
          </div>
</li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Frame
    contextTypes={Object {}}
    copyStackTrace={[Function]}
    frame={
        Object {
            "displayName": "renderFoo",
            "frameworkGroupingOn": false,
            "id": 1,
            "library": false,
            "location": Object {
              "line": 10,
            },
            "source": Object {
              "url": "foo-view.js",
            },
            "toggleFrameworkGrouping": [Function],
          }
    }
    hideLocation={false}
    selectFrame={[Function]}
    selectedFrame={
        Object {
            "displayName": "renderFoo",
            "frameworkGroupingOn": false,
            "id": 1,
            "library": false,
            "location": Object {
              "line": 10,
            },
            "source": Object {
              "url": "foo-view.js",
            },
            "toggleFrameworkGrouping": [Function],
          }
    }
    shouldMapDisplayName={true}
/>,
}
`;
